### ISSUE TITLE
check_valid for 2bit quantization

### ISSUE THREAD

**Issue by zju-stu-lizheng** [mobiusml/gemlite#14]
When I executed the validation code below, I found that gemlite does not maintain the accuracy of the 2bit quantization:
```python
import torch 

def check_valid(x, W, quant_linear, tol=1e-3):
    y_ref = torch.matmul(x, W.T)
    # print(y_ref)
    y_q   = quant_linear(x)
    # print(y_q)
    print((y_ref - y_q).abs().mean())
    assert (y_ref - y_q).abs().mean() < tol, 'Outputs do not match'

############################################################################################
from hqq.core.quantize import HQQLinear, BaseQuantizeConfig

in_features, out_features = 128*4, 128*2
#W_nbits, group_size = 8, in_features 
W_nbits, group_size = 2, 128 
#W_nbits, group_size = 2, 128

linear       = torch.nn.Linear(in_features=in_features, out_features=out_features, bias=False, device='cpu')
quant_config = BaseQuantizeConfig(nbits=W_nbits, group_size=group_size, quant_zero=False, quant_scale=False, axis=1)
hqq_layer    = HQQLinear(linear, quant_config=quant_config, compute_dtype=torch.float16, device='cuda:0', del_orig=False) 

orig_shape   = (out_features, in_features)
W            = hqq_layer.dequantize().reshape(orig_shape).to(torch.float16)
############################################################################################

from gemlite.core import GemLiteLinearTriton, DType
gemlite_linear = GemLiteLinearTriton(W_nbits, 
                                    group_size=group_size, 
                                    in_features=in_features, 
                                    out_features=out_features, 
                                    input_dtype=DType.FP16, 
                                    output_dtype=DType.FP16)

W_q           = hqq_layer.unpack(dtype=torch.uint8).view(orig_shape)
scales        = hqq_layer.meta['scale']
zeros         = hqq_layer.meta['zero']
gemlite_linear.pack(W_q, scales, zeros, None)

x = torch.randn((1, in_features), dtype=torch.float16, device='cuda:0')/10.

check_valid(x, W, gemlite_linear)
```

![Image](https://github.com/user-attachments/assets/1b03c160-1de7-426c-ace3-555d7bc9624a)

**Comment by zju-stu-lizheng** [comment]
When use the config with `in_features, out_features = 4096*4, 4096*2`
```
import random
random.seed(42)
x = torch.randn((1, in_features), dtype=torch.float16, device='cuda:0')/10.

check_valid(x, W, gemlite_linear)
```
the error is tensor(0.0500, device='cuda:0', dtype=torch.float16)

**Comment by mobicham** [comment]
@zju-stu-lizheng thanks for reporting, I can reproduce the result. It seems to me that it's related to autotuning, since the problem only happens with the GEMV_REVSPLITK kernel (batch-size=1) with autotuning, and not the others. If I disable autotuning (just for testing) the output is correct. The strange thing is that 4-bit and 1-bit work fine, I will need to investigate: 

```Python
import torch 

def check_valid(x, W, quant_linear, tol=1e-3):
    y_ref = torch.matmul(x, W.T)
    # print(y_ref)
    y_q   = quant_linear(x)
    # print(y_q)
    print((y_ref - y_q).abs().mean())
    assert (y_ref - y_q).abs().mean() < tol, 'Outputs do not match'

############################################################################################
from hqq.core.quantize import HQQLinear, BaseQuantizeConfig

#in_features, out_features = 128*4, 128*2
#in_features, out_features = 4096, 4096
in_features, out_features = 4096*4, 4096*2

#W_nbits, group_size = 8, in_features 
#W_nbits, group_size = 4, 128 
W_nbits, group_size = 2, 128 
#W_nbits, group_size = 1, 128

linear       = torch.nn.Linear(in_features=in_features, out_features=out_features, bias=False, device='cpu')
quant_config = BaseQuantizeConfig(nbits=W_nbits, group_size=group_size, quant_zero=False, quant_scale=False, axis=1)
hqq_layer    = HQQLinear(linear, quant_config=quant_config, compute_dtype=torch.float16, device='cuda:0', del_orig=False) 

orig_shape   = (out_features, in_features)
W            = hqq_layer.dequantize().reshape(orig_shape).to(torch.float16)
############################################################################################

#For fast checking
from gemlite.core import GemLiteLinearTriton, DType, set_autotune
set_autotune({'GEMV_REVSPLITK':False, 'GEMV_SPLITK': False, 'GEMV':False, 'GEMM_SPLITK':True, 'GEMM':True}, exhaustive=False, use_cuda_graph=False)

from gemlite.core import GemLiteLinearTriton, DType
gemlite_linear = GemLiteLinearTriton(W_nbits, 
                                    group_size=group_size, 
                                    in_features=in_features, 
                                    out_features=out_features, 
                                    input_dtype=DType.FP16, 
                                    output_dtype=DType.FP16,
                                    )

W_q           = hqq_layer.unpack(dtype=torch.uint8).view(orig_shape)
scales        = hqq_layer.meta['scale']
zeros         = hqq_layer.meta['zero']
gemlite_linear.pack(W_q, scales, zeros, None)

#GEMV_REVSPLITK (no autotuning)
torch.manual_seed(42)
x = torch.randn((1, in_features), dtype=torch.float16, device='cuda:0')/10.
check_valid(x, W, gemlite_linear) #0.0001

#GEMM_SPLITK (with autotuning)
torch.manual_seed(42)
x = torch.randn((2, in_features), dtype=torch.float16, device='cuda:0')/10.
check_valid(x, W, gemlite_linear) #0.0001
```

**Comment by mobicham** [comment]
I found the issue and fixed it: https://github.com/mobiusml/gemlite/commit/1774c14a9becfc6cbdcbe0468a67066d1934bd15
You can test with the master branch: `pip install git+https://github.com/mobiusml/gemlite.git`

**Comment by zju-stu-lizheng** [comment]
Thanks, i will try it later.

**Comment by mobicham** [comment]
Actually it's not solved, I still get a problem with `W_bits=1, group_size=32`, the rest looks good. 
If you really want to avoid issues, you can switch to the `GEMV` kernel:
```Python
GemLiteLinearTriton.get_default_gemv = lamda *args, **kwargs: 'GEMV'
```
While I figure out what's going

**Comment by mobicham** [comment]
So, the issue is related to 32-bit packing and impacts only the `GEMV_REVSPLITK` kernel: https://github.com/mobiusml/gemlite/commit/29d67428f54fb80f1e56fb2a8ac6ef1ba9f6b220

It should work with all combinations except 1-bit with a group-size of 32.  To run 1-bit with a group-size 32, there ara two solutions:
1) Use 8-bit packing instead of 32: 
```Python
gemlite_linear.pack(W_q, scales, zeros, None, packing_bitwidth=8) #default is 32
```
2) Revert to the `GEMV` kernel as the default kernel:
```Python
GemLiteLinearTriton.get_default_gemv = lamda *args, **kwargs: 'GEMV'
```

**Comment by zju-stu-lizheng** [comment]
a typo : (lamda -> lambda) `GemLiteLinearTriton.get_default_gemv = lambda *args, **kwargs: 'GEMV'`
